pkg/storage/mvcc.go

```diff
@@ -4170,6 +4176,18 @@ func MVCCGarbageCollect(
 				}
 				count++
 			}
+		} else if meta.Deleted && implicitMeta && !realKeyChanged.IsEmpty() {
```

Comment:
I believe this will incorrectly consider certain keys to be GCable. Consider the case where there is an MVCC range tombstone at 5, a live key at 1, and the GC threshold at 3. We can't GC this until the threshold advances to 5, because a reader at timestamp 4 still needs to see the key. We should have tests that cover this, and checks that guard against it.

Furthermore, there are cases where the above branch will incorrectly use `meta.Timestamp` to adjust stats, specifically when `meta.Timestamp` is an MVCC range tombstone and it falls below `gcKey.Timestamp`. One might argue that the caller sent the wrong `gcKey` timestamp in that case, but we should be resistant to it nonetheless, preferably by doing the correct thing or alternatively erroring out.

Also, `realKeyChanged.IsEmpty()` can happen in two cases: either we're on a bare range key, in which case there is nothing to GC so we should check this and continue right after `mvccGetMetadata()`, or we're on an inline value, in which case `!implicitMeta` should already have us covered.

Reply:
> I believe this will incorrectly consider certain keys to be GCable. Consider the case where there is an MVCC range tombstone at 5, a live key at 1, and the GC threshold at 3. We can't GC this until the threshold advances to 5, because a reader at timestamp 4 still needs to see the key. We should have tests that cover this, and checks that guard against it.

I don't think it is different from the case where we have a point tombstone at t=5 and live key at 1 and GC threshold at 3. If we send request, data at 1 will be deleted.
You can add this as a test with current GC:
https://gist.github.com/aliher1911/7ac4c7e1451952f39248224211919a92
see lines with `!!!!` in comments for this case.
We only do sanity check for removal of naked version and for removal of intents, but all the history analysis is done inside the GC.

We can add more validation here, but not sure it is justified.

> Furthermore, there are cases where the above branch will incorrectly use meta.Timestamp to adjust stats, specifically when meta.Timestamp is an MVCC range tombstone and it falls below gcKey.Timestamp.

That's true, the adjustment is the same though, I think that is the case where we want adjustment to be in one place, but we have extra case which is not covered by existing condition and it is handled specifically. Unless I'm missing something and we do adjustment where it is not supposed to happen. It should be easy to add a use case in gc_test with any history sequence if that is the case.

Reply:
>We only do sanity check for removal of naked version and for removal of intents, but all the history analysis is done inside the GC.
>
>We can add more validation here, but not sure it is justified.

I see, makes sense.

> Unless I'm missing something and we do adjustment where it is not supposed to happen.

I think this would be the case where we have e.g. `b@3`, `[a-c)@4`, and the caller sent a GC key of `b@5`. In this case, the synthetic meta for `b@4` would be used for the stats adjustment, while the key we actually GCed was `b@3`. This could e.g. happen if there was some sort of race where someone else removed `b@5` before the GC request is processed, such as a different GC process.